In [17]:
import numpy as np
from openfermion import InteractionOperator, count_qubits
from openfermion.transforms import get_fermion_operator
from openfermion.linalg import eigenspectrum
from sdstate import *
import pickle

def load_hamiltonain(file_name):
    with open("./Hamiltonians/"+file_name, 'rb') as f:
        loaded_Hf = pickle.load(f)
    return loaded_Hf

if __name__ == "__main__":
#     Steps to run in Lanczos iteration. Recommand to use no more than 3 for large systems of ~100 qubits
    steps = 20
#     Running multiprocessing of Hartree-Fock estimation. Running N threads in parallel for an n-qubit system.
    parallelization = True
#     Running FCI
    FCI = True
    
    #     "LiH", "H4-line", "H4-rectangle", "H4-square", "H2O"
    name = "H2O"
    Hf = load_hamiltonain(name + ".pkl")
    n = count_qubits(Hf)
#     Hf -= of.FermionOperator((), Hf.terms[()])
    print("n_qubit = {}".format(n))
    max_state, min_state, HF_max, HF_min = HF_spectrum_range(Hf, multiprocessing = parallelization)
    HF_dE = HF_max - HF_min
    print("Hartree-Fock dE = {}".format(HF_dE))
    print("Hartree-Fock maximum Slater Determinant state: {}".format(max_state[::-1]))
    print("Hartree-Fock minimum Slater Determinant state: {}".format(min_state[::-1]))
    lanczos_max, lanczos_min = lanczos_total_range(Hf, steps = steps, states = [max_state, min_state])
    dE_lanczos = lanczos_max-lanczos_min
    print(lanczos_max,lanczos_min)
    print("Lanczos algorithm with {} iterations, dE = {}".format(steps, lanczos_dE))
    if FCI:
            eigenvalues = eigenspectrum(Hf, n)
            E_max = max(eigenvalues)
            E_min = min(eigenvalues)
            print(f"FCI maximum: {E_max}, FCI minimum: {E_min}")
            dE_FCI = E_max - E_min
            print(f"Relative Difference: {dE_lanczos/dE_FCI * 100}%")

n_qubit = 14
HF E_max: 17.589436842216198
HF E_min: -48.25730000397694
Hartree-Fock dE = 65.84673684619314
Hartree-Fock maximum Slater Determinant state: 00000000000000
Hartree-Fock minimum Slater Determinant state: 01111111111111
17.589436842216198 -68.14605477365004
Lanczos algorithm with 20 iterations, dE = 13.130453936152458
FCI maximum: 17.589436842216198, FCI minimum: -73.87687308829703
Relative Difference: 93.73450364511186%


n_qubit = 12
HF E_max: 15.8753163276
HF E_min: 12.824966791534049
Hartree-Fock dE = 3.0503495360659514
Hartree-Fock maximum Slater Determinant state: 000000000000
Hartree-Fock minimum Slater Determinant state: 000000011111
FCI maximum: 16.016892019029257, FCI minimum: 2.669082706205123
15.8753163276 2.744862391447542
Lanczos algorithm with 10 iterations, dE = 13.130453936152458

n_qubit = 12
HF E_max: 15.8753163276
HF E_min: 12.824966791534049
Hartree-Fock dE = 3.0503495360659514
Hartree-Fock maximum Slater Determinant state: 000000000000
Hartree-Fock minimum Slater Determinant state: 000000011111
FCI maximum: 16.016892019029257, FCI minimum: 2.669082706205123
98.57716253245923%
15.8753163276 2.7174246467746253
Lanczos algorithm with 30 iterations, dE = 13.130453936152458

n_qubit = 14
HF E_max: 17.589436842216198
HF E_min: -48.25730000397694
Hartree-Fock dE = 65.84673684619314
Hartree-Fock maximum Slater Determinant state: 00000000000000
Hartree-Fock minimum Slater Determinant state: 01111111111111
17.589436842216198 -68.14605477365004
Lanczos algorithm with 20 iterations, dE = 13.130453936152458
FCI maximum: 17.589436842216198, FCI minimum: -73.87687308829703
Relative Difference: 93.73450364511186%1